neo/io/nwbio.py

```diff
@@ -0,0 +1,823 @@
+                                encoded = json.dumps(block.annotations[annotation_name])
+                                annotations[annotation_name].add(encoded)
+                            else:
+                                raise
+                if annotation_name in annotations:
+                    if len(annotations[annotation_name]) > 1:
+                        raise NotImplementedError(
+                            "We don't yet support multiple values for {}".format(annotation_name))
+                    # take single value from set
+                    annotations[annotation_name], = annotations[annotation_name]
+        if "identifier" not in annotations:
+            annotations["identifier"] = self.filename
+        if "session_description" not in annotations:
+            annotations["session_description"] = blocks[0].description or self.filename
+            # todo: concatenate descriptions of multiple blocks if different
+        if "session_start_time" not in annotations:
+            raise Exception("Writing to NWB requires an annotation 'session_start_time'")
+        # todo: handle subject
+        # todo: store additional Neo annotations somewhere in NWB file
+        nwbfile = NWBFile(**annotations)
+
+        assert self.nwb_file_mode in ('w',)  # possibly expand to 'a'ppend later
+        if self.nwb_file_mode == "w" and os.path.exists(self.filename):
+            os.remove(self.filename)
+        io_nwb = pynwb.NWBHDF5IO(self.filename, manager=get_manager(), mode=self.nwb_file_mode)
+
+        if sum(statistics(block)["SpikeTrain"]["count"] for block in blocks) > 0:
+            nwbfile.add_unit_column('_name', 'the name attribute of the SpikeTrain')
+            #nwbfile.add_unit_column('_description', 'the description attribute of the SpikeTrain')
+            nwbfile.add_unit_column(
+                'segment', 'the name of the Neo Segment to which the SpikeTrain belongs')
+            nwbfile.add_unit_column(
+                'block', 'the name of the Neo Block to which the SpikeTrain belongs')
+
+        if sum(statistics(block)["Epoch"]["count"] for block in blocks) > 0:
+            nwbfile.add_epoch_column('_name', 'the name attribute of the Epoch')
+            #nwbfile.add_epoch_column('_description', 'the description attribute of the Epoch')
+            nwbfile.add_epoch_column(
+                'segment', 'the name of the Neo Segment to which the Epoch belongs')
+            nwbfile.add_epoch_column('block', 'the name of the Neo Block to which the Epoch belongs')
+
+        for i, block in enumerate(blocks):
+            self.write_block(nwbfile, block)
+        io_nwb.write(nwbfile)
+        io_nwb.close()
+
+        io_validate = pynwb.NWBHDF5IO(self.filename, "r")
+        errors = pynwb.validate(io_validate, namespace="core")
+        if errors:
+            raise Exception(f"Errors found when validating {self.filename}")
+        io_validate.close()
```

Comment:
```suggestion
        with pynwb.NWBHDF5IO(self.filename, "r") as  io_validate:
            errors = pynwb.validate(io_validate, namespace="core")
            if errors:
                raise Exception(f"Errors found when validating {self.filename}")
```

Using a context is preferred here because it automatically closes the file in case an error prevents the full execution of the inner code.